File: PassArgumentsByValueAndReference/passByValueVsPassByRef.py
"""define two functions and check the pass by value and pass by reference

    functions:

        * func_for_value: a function which takes a integer as its argument
        * func_for_list: a function which takes a list as its argument
"""

import copy

## define functions
def func_for_value(value):
    value = value * 2
    print("value inside func_for_value", value)

def func_for_list_by_copy(lst):
    lst = copy.copy(lst)
    lst.append(42)
    print("list inside func_for_list_by_copy", lst)

def func_for_list(lst):
    lst.append(42)
    print("list inside func_for_list", lst)


## define some example values

num = 10
example_list = [1, 2, 3, 4]


## use the functions for the values

print("before using func_for_value", num)
func_for_value(num)
print("after using func_for_value\n", num)

print("before using func_for_list_by_copy", example_list)
func_for_list_by_copy(example_list)
print("after using func_for_list_by_copy\n", example_list)


print("before using func_for_list", example_list)
func_for_list(example_list)
print("after using func_for_list\n", example_list)

File: PassArgumentsByValueAndReference/passByValueVsPassByRef.c
#include <stdio.h>

// define functions for passing by value and passing by reference

void pass_by_value(int x){
    x = x * 2;
    printf("Inside pass_by_value %d\n", x);
}


void pass_by_reference(int* y){
    *y = *y * 2;
    printf("Inside pass_by_reference %d\n", *y);
}

// use the functions in main function

int main (){
    int example_value = 10;

    printf("initial value of example_value is %d\n", example_value);
    pass_by_value(example_value);
    printf("after calling pass_by_value  example_value is %d\n", example_value);

    printf("before calling pass_by_reference  example_value is %d\n", example_value);
    pass_by_reference(&example_value);
    printf("after calling pass_by_reference  example_value is %d", example_value);

    return 0;
}